content/post/20220520-quantum-compile-part2.md

```diff
@@ -0,0 +1,229 @@
+
+## Intermission
+
+We've seen how some of Coalton's features were used to implement discrete compilation in quilc. Of particular note are type classes and their role in allowing us to define new number types.
+
+However, the tower of abstractions does cause one significant issue: degraded performance. The faster a compiler gets its job done, the better it is for the programmer. In order to make quilc's discrete compilation feature more useful, we had to upgrade Coalton's strategies for optimizing highly polymorphic programs, which we delve into here.
+
+## Efficiency concerns of abstraction + indirection + polymorphism
+
+The correctness of a compiler, and the ease with which one obtains correctness, are only part of the story of practical compiler development. Anybody who has worked on big C++ codebases knows that _compilation speed_ is an enormous practical factor, as well. The speed of quilc is no exception, despite the quality and optimality of its output.
+
+Coalton compiles to Common Lisp. A statically typed language being compiled to a dynamically typed language seems like a raw deal. We went through the effort of ascertaining the type of every term in our program, and we'll just throw it away?
+
+Of course not. Fortunately, Common Lisp allows the programmer to annotate types. The subset of Common Lisp's type system that permits optimizations for speedy code is, for the most part, limited to monomorphic machine, record, and array types. Polymorphism is not expressed in Common Lisp's type system in a way that is practically useful.
+
+Since Coalton knows the type of everything, it can annotate these types in the resulting Common Lisp code. When the code is monomorphic, the Common Lisp compiler will be able to produce fast code.
+
+The story is a little more nuanced with polymorphic code. Coalton supports functions that have either *parametric polymorphism* (one function implementation that supports many types), *ad hoc polymorphism* (a single function name that has many implementations depending on the argument and result types), or both.
+
+A function like `flip` is parametrically polymorphic:
+
+```lisp
+> (type-of 'flip)
+∀ :A :B :C. ((:A → :B → :C) → :B → :A → :C)
+```
+
+Flip can take _any_ two-argument function (taking types `:A` and `:B`), and produce a new function with the arguments flipped (thus taking `:B` and `:A`). It's not too much of a problem compiling this kind of parametric polymorphism in efficient Common Lisp.
+
+Similarly, a function like `+` is ad-hoc-polymorphic:
+
+```lisp
+> (type-of '+)
+∀ :A. NUM :A ⇒ (:A → :A → :A)
+```
+
+This is because how `+` works depends on the type of `:A`. The implementation of `Integer` addition is certainly different from the implementation of `AlgNum` addition. Since we need to be able to work with different implementations of `+` corresponding to different data types at run-time, optimization is more difficult.
```

Comment:
The reference to `AlgNum` is really the only reason one needs to read the prior blog post. If we use a different example (or just reiterate what this type is), then readers can jump into this post first with no problem.

Reply:
deleted the part2 for now